### ISSUE TITLE
Bluebird definitions conflict with lib.es6.d.ts

### ISSUE THREAD

**Issue by cspotcode** [DefinitelyTyped/DefinitelyTyped#4112]
The bluebird definitions attempt to declare a global Promise class.  This conflicts with TypeScript's definitions for ES6's built-in promises, causing compiler errors.

https://github.com/Microsoft/TypeScript/blob/master/bin/lib.es6.d.ts#L4819

https://github.com/borisyankov/DefinitelyTyped/blob/master/bluebird/bluebird.d.ts#L19

**Comment by armorik83** [comment]
Hi.

https://gist.github.com/armorik83/352546c84508a7726d5c
I previously made this patch, but it has a breaking change. Can I PR with this? @vvakame

**Comment by vvakame** [comment]
First. I am not a bluebird user. I say the funniest things possibly.
I think we modified es6-promise/es6-promise.d.ts fit to lib.es6.d.ts(PromiseConstructor and Promise interface).
And new bluebird definition extends it.
I user use `--target es5` with es6-promise.d.ts or use `--target es6` with lib.es6.d.ts.

**Comment by piraveen** [comment]
Here is a working solution (Fix Proposal) => #4788

**Comment by pocesar** [comment]
this must be addressed. I'm having to manually patch the `bluebird.d.ts` file

this also conflicts with `es6-promise.d.ts`

global defined variables and modules must go, for the sake of modularity

**Comment by nevir** [comment]
Running into this as well

**Comment by cspotcode** [comment]
Another thing I noticed: in browser, Bluebird exposes itself as two globals: `Promise` and `P` for short.  This definition only exposes `Promise`.

In the browser, Bluebird overwrites the existing global `Promise` object with itself.  I don't think there's a way for us to do the same in a TypeScript definition.

We could theoretically write a Bluebird definition that *extends* the Promise interfaces found in `lib.es6.d.ts`.  That way, TS's declaration merging would combine the interfaces and would not complain about a name collision.

**Comment by pocesar** [comment]
that makes more sense @cspotcode, since Bluebird is Promises/A+ compliant, and only adds new methods to the Promise object, along with the spec ones.

**Comment by cspotcode** [comment]
If we want to go down this route, I think we'll need two definitions:

* one for node / CommonJS / AMD that declares an external "bluebird" module
* one for browser global that extends the built-in "Promise" interface

Ideally we generate both definitions from a single "source" definition
using some sort of build script.  Does DefinitelyTyped have any precedent
for programmatically generating definitions?  How should we do that?

**Comment by pocesar** [comment]
DT is leaning towards ES6 type modules so that everything will be standardized and future proof. 

Even if the browser uses globals, in the near future we will be either using a transpiler (babel / traceur), a module loader (like System.js) or the support by the browser. 

Globals should be kept at bay because they can break code without you realizing it, including in Node.js (like #4046) 

I myself use Bluebird for everything Promise, be it in the browser or in node.js. If the person wants to expose a global, it should be done by hand IMHO.

Like create a "generic.d.ts" that references `tsd.d.ts` but defines globals and your own modules like:

```typescript
/// <reference path="tsd.d.ts" />

declare var Promise: Bluebird.Class;
declare var _: LodashStatic;
```

This would only be available to the frontend part, and that's an opt-in way to declare globals

**Comment by cspotcode** [comment]
@pocesar: That sounds good in theory, but unfortunately TypeScript will reject it with a type error, because you're trying to re-declare the global `Promise` variable with a different type.  Your suggestion works for Lodash because `_` is a *new* global, not a redefinition of an existing global.  Here is the conflicting type definition: https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es6.d.ts#L5128

That is why I propose creating two type definition files:
* one for the browser global, which extends the interface declared in `lib.es6.d.ts`
* one for ES6 modules / CommonJS / Browserify / Node.js / any other non-global situation, which declares an ES6 module

I agree that we should support *both* use cases in an opt-in way.  The implementation, however, must be different than your suggestion.

**Comment by phenomnomnominal** [comment]
Any progress on this?

**Comment by cspotcode** [comment]
I've started work in a branch: cspotcode/DefinitelyTyped@bluebird-es6-compat

**Comment by StanleyGoldman** [comment]
+1

**Comment by EToreo** [comment]
+1

**Comment by horiuchi** [comment]
+1

**Comment by thynson** [comment]
I find a solution in this way:
bluebird.d.ts
```typescript
declare module __Bluebird {
    class Promise<R> {/*...*/};
    module Promise{/*...*/};
    export = Promise;
}
declare module "bluebird" {
    export = __Bluebird;
}
```
bluebird-global.d.ts
```typescript
///<reference="./bluebird.d.ts"/>
import Promise = __Bluebird.Promise;
```
Is there any problem with this solution?  It pass the test with TypeScript 1.6.2.

**Comment by falsandtru** [comment]
Hi, I made #7305 . Please test on your project.

**Comment by Joshua-F** [comment]
This seems to be fixed with #7321.

**Comment by vvakame** [comment]
:+1:
